nvflare/edge/assessors/model_update.py

```diff
@@ -53,10 +56,132 @@ def __init__(
+                        f"Timeout in {remaining_time:.1f} seconds.",
+                    )
+                    self._last_status_log_time = current_time
+
+    def _start_timeout_tracking(self, fl_ctx: FLContext):
+        """Start independent timeout tracking using thread pool."""
+        if self.timeout_future is not None and not self.timeout_future.done():
+            # Cancel existing timeout if still running
+            self.timeout_future.cancel()
+
+        # Submit new timeout task to thread pool
+        self.timeout_future = self.thread_pool.submit(self._timeout_tracker, fl_ctx)
+        self.log_debug(fl_ctx, f"Started device wait timeout tracking for {self.device_wait_timeout}s")
+
+    def _stop_timeout_tracking(self):
+        """Stop timeout tracking."""
+        if self.timeout_future is not None and not self.timeout_future.done():
+            self.timeout_future.cancel()
+            self.timeout_future = None
+
+    def _timeout_tracker(self, fl_ctx: FLContext):
+        """Independent timeout tracker that runs in thread pool."""
+        try:
+            # Periodic logging during countdown to keep users informed
+            check_interval = 10.0  # Log every 10 seconds
+            elapsed = 0
+
+            while (
+                self.device_wait_start_time is not None
+                and not self.should_stop_job
+                and elapsed < self.device_wait_timeout
+            ):
+
+                time.sleep(min(check_interval, self.device_wait_timeout - elapsed))
+                elapsed += check_interval
+
+                # Check if we should stop monitoring
+                if self.device_wait_start_time is None or self.should_stop_job:
+                    return
+
+                # Log periodic status update
+                if elapsed < self.device_wait_timeout:
+                    remaining = self.device_wait_timeout - elapsed
+                    self.log_info(fl_ctx, f"Device wait countdown: {remaining:.1f} seconds remaining")
+
+            # Check if we're still waiting for devices and timeout exceeded
+            with self.update_lock:
+                if (
+                    self.device_wait_start_time is not None
+                    and not self.should_stop_job
+                    and self._check_device_timeout(fl_ctx)
+                ):
+
+                    # Timeout exceeded, set the stop flag
+                    self.should_stop_job = True
+                    self.log_error(
+                        fl_ctx,
+                        f"Device wait timeout ({self.device_wait_timeout}s) exceeded. "
+                        f"Setting stop job flag to terminate workflow.",
+                    )
+
+        except Exception as e:
+            # Log error but don't crash the thread
+            self.log_error(fl_ctx, f"Error in timeout tracker: {e}")
+        finally:
+            # Clean up
+            self.timeout_future = None
```

Comment:
this is changed both in main thread and this worker thread.

Reply:
updated